### DISCUSSION TITLE
Question fuzzy search in tag pattern filtering

### DISCUSSION THREAD

**Discussion by racng** [milaboratory/mixcr#1059] [General]
According to the Barcode pattern syntax documentation about fuzzy match, indels are not supported. Could this cause over-filtering of reads during alignment step when I include long linker sequences in the tag pattern? We expect indel rate of 1-4% in nanopore sequencing data so I imagine a lot of reads get removed. I see that the refineTagAndSort have parameters to control the amount of indels allowed for correcting Capture Groups. I mistakenly assumed fuzzy matching during align step would also allow indels. Some fuzzy matching methods based on edit distance usually allow deletions, insertions, or substitutions. What is the fuzzy matching done here?

> Uppercase and lowercase letters are used to specify the sequence that must be matched. Capital letters imply perfect match. Lowercase letters allow fuzzy match, where max total mismatches is determined by --tag-max-error-budget (indels are not supported). The --tag-max-error-budget value (default 10) is defined as a total penalty score in bits: - one mismatch with normal nucleotide (a,t,g,c) costs 2 bits, - one mismatch with IUPAC wildcards ((N, w, s, m, etc)) costs 1 bit.

**Comment by PoslavskySV** [comment]
Hi Rachel,

You are right: MiXCR supports only substitutions in tag pattern fuzzy matching. For short patterns you can simulate indels with logical OR, however in your case of long linker sequence in ONT data, indeed you can loose some sequences due to indels. We plan to add support for indels for pattern matching in future, but this is very complicated from the algorithmic point of view, so can't commit for any particular time. On the other hand, MiXCR handles indels in tag refinement step as well as substitutions. Just to check whether  there is a room for improvement, could you share the tag pattern that you use?

**Comment by racng** [comment]
Thanks for the clarification! I can't share the exact linker sequences here so I will represent them as `L{20}` in the tag pattern:
`(CELLBC1:N{10})L{20}(CELLBC2:N{10})L{20}(CELLBC3:N{10})L{20}(MIUMI:aaacaaac)(R1:*)(MIUP3:tgctgaggcttcctgta)`. 

We used a V gene primer that overlaps with `MIUP3`, a constant region added with PCR. Previously I relied on MiXCR to trim off this region, so that I can align with the `--floating-left-alignment-boundary` option. To get more reads that potentially have indels in `MIUP3`, should I just remove that capture group from the tag pattern and then use `--rigid-left-alignment-boundary`? Similarly we used C gene primer that overlaps with `MIUMI`, should I remove that tag group and specify `--rigid-right-alignment-boundary C`?

For the long linker sequences in between CELL barcodes, I am thinking of converting some part of them to `N{9:11}` to allow some amount of indel. For example, `L{20} -> L{5}N{9:11}L{5}`. Or maximize the sensitivity by not looking for the sequence at all, `L{20} -> N{20}`. It would be amazing if MiXCR will support indels for pattern matching in the future so that we don't have to sacrifice too much specificity for higher sensitivity. I do appreciate the amazing work that was already put into making MiXCR such a versatile tool.

I noticed that tag patterns work and even matches a few more reads if I don't put `^` at the beginning. Does MiXCR allow tag pattern search without anchoring the pattern to the beginning of the reads? It is nice if this is an option. Documentation says that `^` is required

> Each pattern has to start with ^ which defines the read beginning. Additionally, one can use $ to specify the end of the read.
